tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1178,187 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
+    *# (if any) and cipher suite.
+    **/
+    if (conn->psk_params.chosen_psk != NULL) {
+        if (potential_match->prf_alg != conn->psk_params.chosen_psk->hmac_alg) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * Used by the server to select a mutually-supported cipher suite.
+ * This function iterates over the server's cipher suite preference list, choosing the first 
+ * suite that is also supported by the client. If an in-group deliminator is encountered, for 
+ * each in-group cipher suite, we iterate through the client's preferences and save the 
+ * index of the client's cipher suite, throwing out the previous index if it is 
+ * less-preferred by the client than the current index. Once the end-group deliminator is 
+ * encountered, the server either accepts the idenfitied cipher suite or if no cipher suite 
+ * has been found, continues iterating through its list. This function checks that the potential cipher
+ * suite match is valid and can be used for the connection.
+ * 
+ */
+static S2N_RESULT s2n_get_mutually_supported_cipher(struct s2n_connection* conn,
+                                                    const uint8_t* wire,
+                                                    uint32_t count,
+                                                    uint32_t cipher_suite_len,
+                                                    struct s2n_cipher_suite** negotiated_cipher) {
+    RESULT_ENSURE_REF(wire);
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(negotiated_cipher);
+    RESULT_ENSURE_EQ(*negotiated_cipher, NULL);
+
+    const struct s2n_security_policy *security_policy;
+    RESULT_GUARD_POSIX(s2n_connection_get_security_policy(conn, &security_policy));
+    const struct s2n_cipher_preferences* cipher_preferences = security_policy->cipher_preferences;
+    RESULT_ENSURE_REF(cipher_preferences);
+
+    bool in_group = false;
+    /** 
+     * Tracks mutually supported cipher suite in the client index. This is needed to track tie-breakers
+     * in an equal-preference group.
+     */
+    int negotiated_client_index = count;
+    /** 
+     * The highest protocol version match index (also a server index). This is a fall back index if
+     * no other mutually-supported cipher suite is identified.
+     */
+    int highest_vers_match_index = -1;
+
+    for (uint32_t i = 0; i < cipher_preferences->count; i++) {
+        struct s2n_cipher_suite* server_cipher = cipher_preferences->suites[i];
+        RESULT_ENSURE_REF(server_cipher);
+
+        /* Check if the cipher suite is an equal-preference grouping delimiter */
+        if (server_cipher == &s2n_equal_preference_group_start) {
+            in_group = true;
+            continue;
+        } 
+        if (server_cipher == &s2n_equal_preference_group_end) {
+            in_group = false;
+            /* Exiting a group and a negotiated cipher has already been found. */
+            if (*negotiated_cipher) {
```

Comment:
Minor nit: I like to do explicit checks for null in if-statements. 
```suggestion
            if (*negotiated_cipher != NULL) {
```

Reply:
This is a styling concern, I think it's cleaner without the explicit not eq null.